pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -110,15 +117,39 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(env model.Environmen
 					v.RequireTls = route.VirtualHost_EXTERNAL_ONLY
 				}
 			}
-			l := buildListener(opts)
-			listeners = append(listeners, l)
 		case model.ProtocolTCP, model.ProtocolMongo:
+			listenerType = plugin.ListenerTypeTCP
 			// TODO
 			// Look at virtual service specs, and identity destinations,
 			// call buildOutboundNetworkFilters.. and then construct TCPListener
 			//buildTCPListener(buildOutboundNetworkFilters(clusterName, addresses, servicePort),
 			//	listenAddress, uint32(servicePort.Port), servicePort.Protocol)
 		}
+
+		newListener := buildListener(opts)
+
+		params := &plugin.CallbackListenerInputParams{
+			ListenerType: listenerType,
+			Env:          &env,
+			Node:         &node,
+		}
+		mutable := &plugin.CallbackListenerMutableObjects{
+			Listener:    newListener,
+			TCPFilters:  &networkFilters,
+			HTTPFilters: &hTTPFilters,
+		}
+		for _, p := range configgen.Plugins {
+			if err := p.OnOutboundListener(params, mutable); err != nil {
+				log.Error(err.Error())
```

Comment:
`log.Warn`; `log.Error` is for irrecoverable errors that force Pilot to stop serving.

Reply:
done